napari/_vispy/visuals/volume.py

```diff
@@ -34,6 +37,56 @@
     return adjacent_bg;
 }
 
+vec3 calculateGradient(vec3 loc, vec3 step, out int n_bg_borders) {
+    // calculate gradient within the volume by finite differences
+
+    n_bg_borders = 0;
+    vec3 G = vec3(0.0);
+
+    float prev;
+    float next;
+
+    prev = colorToVal($get_data(loc - vec3(step.x, 0, 0)));
+    next = colorToVal($get_data(loc + vec3(step.x, 0, 0)));
+    n_bg_borders += detectAdjacentBackground(prev, next);
+    G.x = next - prev;
+
+    prev = colorToVal($get_data(loc - vec3(0, step.y, 0)));
+    next = colorToVal($get_data(loc + vec3(0, step.y, 0)));
+    n_bg_borders += detectAdjacentBackground(prev, next);
+    G.y = next - prev;
+
+    prev = colorToVal($get_data(loc - vec3(0, 0, step.z)));
+    next = colorToVal($get_data(loc + vec3(0, 0, step.z)));
+    n_bg_borders += detectAdjacentBackground(prev, next);
+    G.z = next - prev;
+
+    return G;
+}
+
+vec3 calculateIsotropicGradient(vec3 loc, vec3 step) {
+    // calculate gradient within the volume by finite differences
+    // using a 3D sobel-feldman kernel
+
+    vec3 G = vec3(0.0);
+
+    for (int i=-1; i <= 1; i++) {
+        for (int j=-1; j <= 1; j++) {
+            for (int k=-1; k <= 1; k++) {
+                float val = colorToVal($get_data(loc + vec3(i, j, k) * step));
```

Comment:
I guess this was already the case in the normal gradient but: how are overflow edge cases handled?

Reply:
I just checked the vispy Volume visual and it's set to `clamp_to_edge`.